### PR TITLE
Use a validating component for numbers

### LINKED ISSUE #29: Create a self-contained self-validating number input component
Expected Behavior
A nicely self-contained, reusable component that can clean it's own input

Current Behavior
Logic to clean input is spread across parent component (sendAmount, exchangeAmount)
See #4414

Looks like we already have two components for this, let's audit those see if they can be modified to fit the job:
https://github.com/celo-org/celo-monorepo/blob/master/packages/react-components/components/NumberInput.tsx
https://github.com/celo-org/celo-monorepo/blob/master/packages/react-components/components/PhoneNumberInput.tsx

See here too: https://github.com/celo-org/celo-monorepo/blob/master/packages/mobile/src/send/LabeledTextInput.tsx#L37

### PR DESCRIPTION
**Pull request by drklee3**
### Description

Implemented a `ValidatedTextInput` component which supports integer, decimal, and phone validators.  This prevents users from inputting invalid characters such as letters and multiple decimal points in the `SendAmount` amount field.

The component also allows for a set number of decimal points and prevents further input after reaching such set limit.  It also supports commas as decimal points, based on which language is set. (if lng.startsWith('es') use comma decimal points)

### Tested

Tests pass after various snapshots were updated.  The input validation logic also has a number of tests created to format integers decimals, and phone numbers.

### Other changes

* Removed the `NumberInput` component, replaced with inline TextInput components
* Show SendAmount invalid amount error instead of re-focusing amount input (error did not show up before previously)
* Various prop names were renamed to be more consistent

### Related issues

- Fixes #29

### CODE REVIEW

File: packages/react-components/components/ValidatedTextInput.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+// Required props when validator type is phone
+interface PhoneValidatorProps {
+  validator: 'phone'
+  countryCallingCode: string
+  // Following props unused w/ 'phone' validator but required to be defined
+  customValidator?: CustomValidatorProps['customValidator']
+}
+
+interface NumberValidatorProps {
+  validator?: 'integer' | 'decimal'
```
Comment (jmrossy) [inline packages/react-components/components/ValidatedTextInput.tsx]: A common pattern in TS is to use a `kind` property to differentiate interfaces like these:
https://github.com/celo-org/celo-monorepo/blob/master/packages/mobile/src/utils/recipient.ts#L9

File: packages/utils/test/inputValidation.test.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+describe('inputValidation', () => {
+  function validateFunction(
+    desc: string,
+    itStr: string,
+    inputs: string[],
+    validator: 'phone' | 'integer' | 'decimal',
+    expected: string,
+    props?: BaseProps
+  ) {
+    describe(desc, () => {
```
Comment (jmrossy) [inline packages/utils/test/inputValidation.test.ts]: Typically in a test file there's a root `describe` wrapper with children `it` wrappers:
https://github.com/celo-org/celo-monorepo/blob/master/packages/utils/test/parsing.test.ts

File: packages/mobile/src/pincode/Pincode.tsx [hunk 3]
```diff
@@ -163,31 +162,35 @@ export class Pincode extends React.Component<Props, State> {
         return (
           <View style={style.pincodeContent}>
             <Text style={[fontStyles.h1, style.h1]}>{t('createPin.title')}</Text>
-            <NumberInput
```
Comment (jmrossy) [inline packages/mobile/src/pincode/Pincode.tsx:166]: Why change this?

**Review by jmrossy — COMMENTED** [review]
Overall I think your approach here looks good. Please sort out the failing tests and then I'll have a closer look :)

File: packages/mobile/src/pincode/Pincode.tsx (hunk 3 above)
Comment (drklee3) [reply packages/mobile/src/pincode/Pincode.tsx:166]: Originally I was going to merge `NumberInput` into the `ValidatedTextInput` but it seemed like the component wasn't necessary in the first place so I changed it to use `TextInput`.  I guess the change isn't really directly related anymore so I can revert this change if needed.

Comment (jmrossy) [reply packages/mobile/src/pincode/Pincode.tsx:166]: The pin is meant to be a number. Unless you tested this and found it works better, seems like some kind of NumberInput is the better here

File: packages/utils/src/inputValidation.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+export function validateInteger(input: string): string {
+  return input.replace(/[^0-9]/g, '')
+}
+
+export function validateDecimal(input: string, lng?: string): string {
+  // Comma decimal points, only supports es right now!
+  if (lng && lng.startsWith('es')) {
+    // Keep only first decimal
+    return input
+      .replace(/[^0-9,]/g, '')
```
Comment (jmrossy) [inline packages/utils/src/inputValidation.ts]: Can this be achieved without 4 consecutive regex replacements?

**Review by jmrossy — COMMENTED** [review]
Looks good, almost ready!

With this change, I guess the hierarchy of components is:
`ValidatedTextInput -> LabeledTextInput -> TextInput (custom) -> TextInput (default)`
Is that correct? If so, can you add a comment at the top of each file explaining when the expected usage scenario is for these components :)

File: packages/mobile/src/pincode/Pincode.tsx (hunk 3 above)
Comment (drklee3) [reply packages/mobile/src/pincode/Pincode.tsx:166]: From what I've found, the `NumberInput` component didn't have any extra functionality other than having `maxLength={6}` and wrapped in `View` with some padding.

While it uses the numeric keyboardType which only allows numbers to be input, it's still possible to input invalid characters via copy-pasting so I've just now replaced it with the `ValidatedTextInput`.

**Review by jmrossy — APPROVED** [review]

**Pull request merged**